vertical-pod-autoscaler/recommender/util/circularbuffer.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+// CircularBuffer is a queue with a fixed maximum capacity. Elements are
+// inserted/removed in the FIFO order. Elements are removed from the buffer
+// when it runs out of capacity and a new element is inserted.
+type CircularBuffer interface {
+	// Add a value to the end of the queue. On overflow returns true and the
+	// oldest value, which is also removed from the buffer. Otherwise
+	// returns (false, _).
+	Push(value float64) (bool, float64)
+
+	// Returns the elements in the buffer, ordered by time of insertion
+	// (oldest first).
+	Contents() []float64
+
+	// Returns a pointer to the most recently added element. The pointer can
+	// be used to modify the last element. It is only valid until the next
+	// call to Push(). Return nil if called on an empty buffer.
+	Head() *float64
```

Comment:
I think it's less relevant now as it's clear this is used as a sliding window (with given length).
In any case, if someone wants to check the current number of elements, they can call len(x.Contents()).